aten/src/ATen/native/cuda/FusedAdamKernel.cu

```diff
@@ -0,0 +1,228 @@
+#define TORCH_ASSERT_ONLY_METHOD_OPERATORS
+#include <ATen/Dispatch.h>
+#include <ATen/native/ForeachUtils.h>
+#include <ATen/native/cuda/ForeachFunctors.cuh>
+#include <ATen/native/cuda/MultiTensorApply.cuh>
+#include <c10/macros/Macros.h>
+#include <c10/util/irange.h>
+#include <vector>
+
+
+namespace at { namespace native {
+
+namespace {
+constexpr uint8_t kParamIdx = 0;
+constexpr uint8_t kGradIdx = 1;
+constexpr uint8_t kExpAvgIdx = 2;
+constexpr uint8_t kExpAvgSqIdx = 3;
+constexpr uint8_t kMaxExpAvgSqIdx = 4;
+
+template <typename scalar_type, typename opmath_t, int depth=4>
+C10_DEVICE __forceinline__ void adam_math(
+    scalar_type r_args[depth][kILP],
+    const float* step_count,
+    const double lr,
+    const double beta1,
+    const double beta2,
+    const double weight_decay,
+    const double eps,
+    const bool maximize,
+    const bool amsgrad,
+    const float* grad_scale_ptr,
+    const float* found_inf_ptr
+) {
+#pragma unroll
+    for (int ii = 0; ii < kILP; ii++) {
+        // Load values.
+        opmath_t param = static_cast<opmath_t>(r_args[kParamIdx][ii]);
+        opmath_t grad = static_cast<opmath_t>(r_args[kGradIdx][ii]);
+        if (grad_scale_ptr) {
+            grad /= (static_cast<double>(*grad_scale_ptr));
+        }
+        const opmath_t grad_to_store = grad;
+        if (maximize) {
+            grad = -grad;
+        }
+        opmath_t exp_avg = static_cast<opmath_t>(r_args[kExpAvgIdx][ii]);
+        opmath_t exp_avg_sq = static_cast<opmath_t>(r_args[kExpAvgSqIdx][ii]);
+        opmath_t max_exp_avg_sq;
+        if (amsgrad) {
+            max_exp_avg_sq = static_cast<opmath_t>(r_args[kMaxExpAvgSqIdx][ii]);
+        }
+
+        // Update param, grad, 1st and 2nd order momentum.
+        if (weight_decay != 0) {
+            grad += param * weight_decay;
+        }
+        exp_avg = beta1 * exp_avg + (1 - beta1) * grad;
+        exp_avg_sq = beta2 * exp_avg_sq + (1 - beta2) * grad * grad;
+
+        if (amsgrad) {
+            max_exp_avg_sq = ::max(max_exp_avg_sq, exp_avg_sq);
+        }
+
+        const opmath_t bias_correction1 = 1 - ::pow(beta1, *step_count);
+        const opmath_t bias_correction2 = 1 - ::pow(beta2, *step_count);
+
+        const opmath_t step_size = lr / bias_correction1;
+
+        const opmath_t bias_correction2_sqrt = ::sqrt(bias_correction2);
+
+        opmath_t denom;
+        if (amsgrad) {
+            denom = (::sqrt(max_exp_avg_sq) / bias_correction2_sqrt) + eps;
+        } else {
+            denom = (::sqrt(exp_avg_sq) / bias_correction2_sqrt) + eps;
+        }
+
+        param -= step_size * exp_avg / denom;
+
+        // Store results.
+        r_args[kParamIdx][ii] = param;
+        r_args[kGradIdx][ii] = grad_to_store;
```

Comment:
why are you unconditionally storing grad even if it wasn't scaled?

Reply:
good catch. No reasons, I forgot to update here.